### PR TITLE
Update CA1710 to allow empty suffix

### PR DESCRIPTION
**Pull request by Evangelink**
Contributes to #1818 and allows for empty suffix as a way to override some of the built-in suffixes.

### CODE REVIEW

File: docs/Analyzer Configuration.md [hunk 1]
```diff
@@ -452,5 +452,6 @@ Examples:
 
 | Option Value | Summary |
 | --- | --- |
-|`dotnet_code_quality.CA1710.additional_required_suffixes = MyClass->Class` | All types inheriting from MyClass are expected to have the 'Class' suffix |
-|`dotnet_code_quality.CA1710.additional_required_suffixes = MyClass->Class|MyNamespace.IPath->Path` | All types inheriting from 'MyClass' are expected to have the 'Class' suffix AND all types implementing 'MyNamespace.IPath' are expected to have the 'Path' suffix. |
\ No newline at end of file
+|`dotnet_code_quality.CA1710.additional_required_suffixes = MyClass->Class` | All types inheriting from' MyClass' are expected to have the 'Class' suffix |
```
Comment (Evangelink) [inline docs/Analyzer Configuration.md]: I have added single quotes around `MyClass`.

File: docs/Analyzer Configuration.md [hunk 2]
```diff
@@ -452,5 +452,6 @@ Examples:
...
+|`dotnet_code_quality.CA1710.additional_required_suffixes = MyClass->Class|MyNamespace.IPath->Path` | All types inheriting from 'MyClass' are expected to have the 'Class' suffix AND all types implementing 'MyNamespace.IPath' are expected to have the 'Path' suffix. |
+|`dotnet_code_quality.CA1710.additional_required_suffixes = T:System.Data.IDataReader->` | Allows to override built-in suffixes, in this case, all types implementing 'IDataReader' are no longer expected to end in 'Collection' |
```
Comment (mavasani) [inline docs/Analyzer Configuration.md]: I think we should not require the input to have a trailing `->`, so even the below setting should produce the same result.
```suggestion
|`dotnet_code_quality.CA1710.additional_required_suffixes = T:System.Data.IDataReader` | Allows to override built-in suffixes, in this case, all types implementing 'IDataReader' are no longer expected to end in 'Collection' |
```

Comment (mavasani) [reply docs/Analyzer Configuration.md]: Or maybe for clarity we can allow some special character after the `->`, say `_` or some such?

Comment (Evangelink) [reply docs/Analyzer Configuration.md]: I like the second option.

Comment (Evangelink) [reply docs/Analyzer Configuration.md]: The symbol `_` was a good way to express discard but the problem is that it is a valid class name suffix (in term of what is allowed) even if it's not a nice suffix.

File: docs/Analyzer Configuration.md (hunk 1 above)
Comment (mavasani) [inline docs/Analyzer Configuration.md]:
```suggestion
|`dotnet_code_quality.CA1710.additional_required_suffixes = MyClass->Class` | All types inheriting from 'MyClass' are expected to have the 'Class' suffix |
```

File: src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/IdentifiersShouldHaveCorrectSuffixTests.cs [hunk 3]
```diff
@@ -1380,6 +1380,261 @@ Inherits Dictionary(Of String, String)
             }.RunAsync();
         }
 
+        [Fact, WorkItem(1818, "https://github.com/dotnet/roslyn-analyzers/issues/1818")]
+        public async Task CA1710_AllowEmptySuffix()
+        {
+            var editorConfigText = @"dotnet_code_quality.CA1710.additional_required_suffixes = T:System.Data.IDataReader->{}";
```
Comment (mavasani) [inline src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/IdentifiersShouldHaveCorrectSuffixTests.cs]: Can you make this a Theory based test and also verify the behavior for different use of whitespaces, say `T:System.Data.IDataReader -> {}` and `T:System.Data.IDataReader-> { } `?

File: src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/IdentifiersShouldHaveCorrectSuffixTests.cs [hunk 4]
```diff
@@ -1452,6 +1452,283 @@ End Sub
...
+        // Following values are ok
+        [InlineData("dotnet_code_quality.CA1710.additional_required_suffixes = T:System.Data.IDataReader->{}")]
+        [InlineData("dotnet_code_quality.CA1710.additional_required_suffixes = T:System.Data.IDataReader-> {}")]
+        [InlineData("dotnet_code_quality.CA1710.additional_required_suffixes = T:System.Data.IDataReader->{} ")]
+        [InlineData("dotnet_code_quality.CA1710.additional_required_suffixes = T:System.Data.IDataReader-> {} ")]
+        // Following values are not ok
+        [InlineData("dotnet_code_quality.CA1710.additional_required_suffixes = T:System.Data.IDataReader->{ }")]
+        [InlineData("dotnet_code_quality.CA1710.additional_required_suffixes = T:System.Data.IDataReader-> { }")]
+        [InlineData("dotnet_code_quality.CA1710.additional_required_suffixes = T:System.Data.IDataReader->{ } ")]
+        [InlineData("dotnet_code_quality.CA1710.additional_required_suffixes = T:System.Data.IDataReader-> { } ")]
```
Comment (Evangelink) [inline src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/IdentifiersShouldHaveCorrectSuffixTests.cs]: Not sure if you were expecting these cases to be supported or not.

Comment (mavasani) [reply src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/IdentifiersShouldHaveCorrectSuffixTests.cs]: All of these values should work. We should trim the leading and trailing whitespaces in each value.

Comment (Evangelink) [reply src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/IdentifiersShouldHaveCorrectSuffixTests.cs]: I'm already doing a trim (start and end). The values I have highlighted here are the ones for which the user has put space between the curly braces. Is it something we want to handle? If so how many spaces shall be allowed?

Comment (mavasani) [reply src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/IdentifiersShouldHaveCorrectSuffixTests.cs]: IMO All spaces should be ignored. I think you can handle trimmed values starting and ending with curliest and any number of spaces in between as empty value.

**Pull request merged**